### PR TITLE
docs: add Colab CLI onboarding guide (install + sanity check + RankZe…

### PR DESCRIPTION
**Pull request by XiandaDu**
# PR description (branch: docs/install-colab-cli)

> expected score (0.6771) assumes the monoT5 attention-mask fix is on main.

# Pull Request Checklist

## Reference Issue

ref: N/A (supersedes #407 / #408 / #409 per professor's feedback)

## Summary

- Adds `docs/install-colab-cli.md`: a lightweight onboarding guide in the style of the anserini/pyserini onboarding sequence — (1) install the [Colab CLI](https://github.com/googlecolab/google-colab-cli) and sanity-check it, (2) run an end-to-end multi-stage retrieval pipeline (BM25 → monoT5-base rerank, TREC DL20) — **entirely on Colab's free tier** (T4).
- Students type each `colab` command themselves (`new` → `exec` / `console` → `download` → `stop`); the core experiment runs line-by-line in an interactive `colab console` shell. Only environment setup is prepackaged (`scripts/colab_setup.py`, fetched with curl and sent via `colab exec -f`).
- Expected result: nDCG@10 0.4796 (BM25) → **0.6771** (monoT5-base rerank) — the multi-stage lift is the pedagogical payoff. Guide ends with a Reproduction Log section like the previous onboarding guides.
- Adds a two-line pointer at the top of `docs/onboarding.md` (intended sequence: `install-colab-cli.md` → `onboarding.md`). A closing section shows how to run the onboarding RankZephyr/FirstMistral experiments the same way once students have Colab Pro (7B models don't fit the free T4).

## Why

Maintainer feedback on the recipes stack (#407–#409) asked for something much more lightweight: a doc students follow and reproduce before onboarding, like the anserini/pyserini sequence, rather than wrapped one-command automation. Free tier keeps the entry cost at zero (Colab Pro only becomes necessary at the onboarding.md step, as that guide already explains). The hard-won details from the recipes (JDK 21, exec timeouts, streaming child-process output, the pyserini import-time `OPENAI_API_KEY` quirk) survive as inline explanations.

## Validation

```bash
.venv/bin/pre-commit run --files docs/install-colab-cli.md scripts/colab_setup.py docs/onboarding.md
python3 -m py_compile scripts/colab_setup.py
```

Live end-to-end pass on an actual free-tier T4 via the Colab CLI (with the attention-mask fix applied):
- `colab new -s ... --gpu T4` → `colab exec -f scripts/colab_setup.py` → `setup ok`
- piped `colab console` smoke test (real shell on the VM confirmed)
- full BM25 → monoT5 → trec_eval pipeline on the VM in 230 s total: nDCG@1/5/10 = 0.8210/0.7045/0.6771, matching the local RTX 4090 run bit-for-bit
- `colab download` of the artifact tarball and `colab stop` cleanup verified
- BM25 baseline re-derived from the retrieval cache: 0.4796

## Checklist Items

- [x] Have you followed the [contributing guidelines](CONTRIBUTING.md)?
- [x] Have you verified that there are no existing [Pull Requests](https://github.com/castorini/rank_llm/pulls) for the same update/change?
- [x] Have you updated any relevant documentation or added new tests where needed?

# PR Type

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no API changes)
- [x] Documentation content changes
- [ ] Reproduction logs
- [ ] Other...

---

# Close comment for #407 (feat/colab-cli-1-common)

Per feedback, replacing this scripts-based stack with a lightweight onboarding doc: #<NEW_PR> adds `docs/install-colab-cli.md` (install + sanity check + a free-tier BM25 → monoT5 repro, in the style of the anserini/pyserini onboarding sequence). The shared plumbing here is preserved on the branch in case we later want one-command automation on top of the doc flow. Closing.

# Close comment for #408 (feat/colab-cli-2-rerank)

Superseded by the lightweight doc approach in #<NEW_PR> — the reranking flow there is the same idea this recipe automated, but as step-by-step commands students run themselves (and sized to the free tier). Branch preserved. Closing.

# Close comment for #409 (feat/colab-cli-3-train)

Superseded by the lightweight doc approach in #<NEW_PR>, which covers the onboarding use case. The training recipe and single-GPU accelerate config are out of scope for that doc; branch preserved if remote training comes back later. Closing.

### CODE REVIEW

File: docs/install-colab-cli.md [hunk 1]
```diff
@@ -0,0 +1,238 @@
+# Install the Colab CLI and Rerank on a Free Colab GPU
+
+The rank_llm step of the onboarding path requires a GPU, and the
```
Comment (lintool) [inline docs/install-colab-cli.md]: btw, our convention is each sentence on its own line.
this makes diffs easier to read - can you please reflow in this way?

**Review by lintool — CHANGES_REQUESTED** [review]

File: scripts/colab_setup.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
+"""One-time environment setup for the docs/install-colab-cli.md guide.
```
Comment (lintool) [inline scripts/colab_setup.py:1]: would it be better to put this in `scripts/`, a generic directory we can throw scripts in?

Comment (XiandaDu) [reply scripts/colab_setup.py:1]: moved there!

File: src/rank_llm/rerank/pairwise/duot5.py [hunk 3]
```diff
@@ -72,7 +72,14 @@ def run_llm_batched(
...
 
-        outputs = self._llm.generate(input_ids, generation_config=gen_cfg)
+        # transformers>=5 only infers a missing attention mask for decoder-only
+        # models, so pass it explicitly; otherwise the encoder attends to the
+        # padding of shorter prompts in the batch and their scores degrade.
+        outputs = self._llm.generate(
+            input_ids,
+            attention_mask=tokenized["attention_mask"],
+            generation_config=gen_cfg,
+        )
```
Comment (XiandaDu) [inline src/rank_llm/rerank/pairwise/duot5.py:82]: monot5.py and duot5.py call generate() with only input_ids, never an attention_mask. They always have — it worked because Transformers built the mask for us. v5 narrowed that: v4 built the mask for these calls (v4.57.1:2454 (https://github.com/huggingface/transformers/blob/v4.57.1/src/transformers/generation/utils.py#L2454)), v5 only does it for decoder-only models (v5.7.0:2426 (https://github.com/huggingface/transformers/blob/v5.7.0/src/transformers/generation/utils.py#L2426)). T5 is encoder-decoder, so it gets nothing — and the encoder now reads the padding of shorter prompts in the batch.

**Comment by lintool** [comment]
hi @XiandaDu can we reconcile with https://aclanthology.org/2020.findings-emnlp.63/ ?

The paper doesn't actually evaluate DL20, but if you run the eval on MS MARCO, you should be able to repro the same numbers? (Or at least similar?)

This is primarily for sanity checking, because asking students to repro on MS MARCO would take too long - DL20 makes sense.

File: src/rank_llm/rerank/pointwise/monot5.py [hunk 4]
```diff
@@ -60,9 +60,15 @@ def run_llm_batched(
...
         ).to(self._device)
 
+        # transformers>=5 only infers a missing attention mask for decoder-only
+        # models, so pass it explicitly; otherwise the encoder attends to the
+        # padding of shorter prompts in the batch and their scores degrade.
+        attention_mask = token_prompts["attention_mask"]
         token_prompts = token_prompts["input_ids"]
 
-        batch_outputs = self._llm.generate(token_prompts, generation_config=gen_cfg)
+        batch_outputs = self._llm.generate(
```
Comment (lintool) [inline src/rank_llm/rerank/pointwise/monot5.py:71]: I think we can reflow for this to be on a single line.

Comment (XiandaDu) [reply src/rank_llm/rerank/pointwise/monot5.py:71]: fixed

**Comment by XiandaDu** [comment]
> hi @XiandaDu can we reconcile with https://aclanthology.org/2020.findings-emnlp.63/ ?
> 
> The paper doesn't actually evaluate DL20, but if you run the eval on MS MARCO, you should be able to repro the same numbers? (Or at least similar?)
> 
> This is primarily for sanity checking, because asking students to repro on MS MARCO would take too long - DL20 makes sense.

Ran it on MS MARCO passage dev, BM25 top-1000, MRR@10, 500-query sample:
|               | paper | ours   |
|---------------|-------|--------|
| BM25          | 0.184 | 0.1952 |
| + monoT5-base | 0.381 | 0.3820 |

I think they match. BM25 is a touch high since the prebuilt index uses different parameters than 2020-era Anserini.

This only reproduces after the attention mask fix. monoT5/duoT5 never passed an attention_mask to generate(), which transformers 4 filled in for us but transformers 5 no longer does for encoder-decoder models. Without the new lines I added, the same run gives 0.3040.

**Review by lintool — APPROVED** [review]
